File: Minimax.py
from Move import Move
from Kalah import Kalah
from Side import Side


class Minimax:

    isPreviousMoveFromLeft = 0
    isPreviousMoveFromRight = 0

    def __init__(self, side):
        self.mySide = side

    def minimax(self, kalah_board, alpha, beta, side, depth):
        board = kalah_board.getBoard()
        if depth == 0 or kalah_board.gameOver():
            return self.evalHeuristics(kalah_board.getBoard())

        if side == self.mySide:
            max_evaluation = - 9999
            self.isPreviousMoveFromLeft = 1
            for possible_move in range(1, 8):
                board_copy = board.copyBoard(board)
                kalah_copy = Kalah(board_copy)
                move = Move(side, possible_move)
                if kalah_copy.isLegalMove(move):
                    self.isPreviousMoveFromRight = self.isMoveFromRightMostHole(board_copy, side, possible_move)
                    nextSide = kalah_copy.makeMove(move)
                    evaluation = self.minimax(kalah_copy, alpha, beta, nextSide, depth - 1)
                    max_evaluation = max(max_evaluation, evaluation)
                    alpha = max(alpha, evaluation)
                    if beta <= alpha:
                        break
                    self.isPreviousMoveFromLeft = 0
            return max_evaluation
        else:
            min_evaluation = 9999
            for possible_move in range(1, 8):
                board_copy = board.copyBoard(board)
                kalah_copy = Kalah(board_copy)
                move = Move(side, possible_move)
                if kalah_copy.isLegalMove(move):
                    nextSide = kalah_copy.makeMove(move)
                    evaluation = self.minimax(kalah_copy, alpha, beta, nextSide, depth - 1)
                    min_evaluation = min(min_evaluation, evaluation)
                    beta = min(beta, evaluation)
                    if beta <= alpha:
                        break
            return min_evaluation

    def evalHeuristics(self, board):
        # Weights depth 9
        w1 = 0.675
        w2 = 0.275
        w3 = 0.675
        w4 = 1
        w5 = 0.5
        w6 = 0.7
        w7 = 0.475

        h1 = self.hoardInLeftmostHole(board, self.mySide)
        h2 = self.hoardOnMySide(board, self.mySide)
        h3 = self.possibleMoves(board, self.mySide)
        h4 = self.seedsInStore(board, self.mySide)
        h5 = self.isPreviousMoveFromRight
        h6 = self.seedsInStore(board, Side.opposite(self.mySide))
        h7 = self.isPreviousMoveFromLeft

        return h1 * w1 + h2 * w2 + h3 * w3 + h4 * w4 + h5 * w5 - h6 * w6 + h7 * w7

    def isMoveFromLeftMostHole(self, board, side, move):
        result = 0

        for i in range(1, 8):
            if board.getSeeds(side, i) != 0:
                if move == i:
                    result = 1
                break
        return result

    def seedsInStore(self, board, side):
        # Calculate a score based on how many seeds are in the store based on the move
        score = board.getSeedsInStore(side)
        return score

    def hoardInLeftmostHole(self, board, side):
        # Calculate the score based on how many seeds there are in the leftmost hole.
        score = board.getSeeds(side, 1)
        return score

    def hoardOnMySide(self, board, side):
        # Calculate the score based on how many seeds there are on my side.
        score = 0

        for i in range(1, 8):
            score += board.getSeeds(side, i)

        return score

    def possibleMoves(self, board, side):
        # Calculate how many holes are not empty
        score = 0

        for i in range(1, 8):
            if board.getSeeds(side, i) != 0:
                score += 1
        return score

    def isMoveFromRightMostHole(self, board, side, move):
        result = 0

        for i in range(7, 0, -1):
            if board.getSeeds(side, i) != 0:
                if move == i:
                    result = 1
                break
        return result

File: javaAgent/src/MiniMax.java
import static java.lang.Double.max;
import static java.lang.Double.min;

public class MiniMax {

    int isPreviousMoveFromRight = 0;

    Side mySide;

    public MiniMax(Side side){
        this.mySide = side;
    }

    double minimax(Kalah kalah_board, double alpha, double beta, Side side, int depth){
        Board board = kalah_board.getBoard();

        if (depth == 0 || kalah_board.gameOver()){
            return evalHeuristics(kalah_board.getBoard());
        }

        if (side == mySide) {
            double max_evaluation = - 9999;

            for (int possible_move = 1; possible_move <= 7; possible_move++){
                Board board_copy = new Board(board);
                Kalah kalah_copy = new Kalah(board_copy);
                Move move = new Move(side, possible_move);
                if (kalah_copy.isLegalMove(move)){
                    isPreviousMoveFromRight = isMoveFromRightMostHole(board_copy, side, possible_move);
                    Side nextSide = kalah_copy.makeMove(move);
                    double evaluation = minimax(kalah_copy, alpha, beta, nextSide, depth - 1);
                    max_evaluation = max(max_evaluation, evaluation);
                    alpha = max(alpha, evaluation);
                    if (beta <= alpha)
                        break;
                }
            }
            return max_evaluation;
        }
        else {
            double min_evaluation = 9999;
            for (int possible_move = 1; possible_move <= 7; possible_move++){
                Board board_copy = new Board(board);
                Kalah kalah_copy = new Kalah(board_copy);
                Move move = new Move(side, possible_move);

                if (kalah_copy.isLegalMove(move)){
                    Side nextSide = kalah_copy.makeMove(move);
                    double evaluation = minimax(kalah_copy, alpha, beta, nextSide, depth - 1);
                    min_evaluation = min(min_evaluation, evaluation);
                    beta = min(beta, evaluation);
                    if (beta <= alpha)
                        break;
                }
            }
            return min_evaluation;
        }
    }


    double evalHeuristics(Board board){
        double w1 = 0.198649;
        double w2 = 0.190084;
        double w3 = 0.370793;
        double w4 = 1;
        double w5 = 0.418841;
        double w6 = 0.565937;

        double h1 = hoardInLeftmostHole(board, mySide);
        double h2 = hoardOnMySide(board, mySide);
        double h3 = possibleMoves(board, mySide);
        double h4 = seedsInStore(board, mySide);
        double h5 = isPreviousMoveFromRight;
        double h6 = seedsInStore(board, mySide.opposite());

        return h1 * w1 + h2 * w2 + h3 * w3 + h4 * w4 + h5 * w5 - h6 * w6;
    }

    int seedsInStore(Board board, Side side){
        return board.getSeedsInStore(side);
    }

    int hoardInLeftmostHole(Board board, Side side){
        return board.getSeeds(side, 1);
    }

    int hoardOnMySide(Board board, Side side){
        int score = 0;

        for (int i = 1; i <= 7; i++){
            score += board.getSeeds(side, i);
        }

        return score;
    }

    int possibleMoves(Board board, Side side){
        int score = 0;

        for (int i = 1; i <= 7; i++){
            if (board.getSeeds(side, i) != 0)
                score += 1;
        }

        return score;
    }

    int isMoveFromRightMostHole(Board board, Side side, int move){
        int result = 0;

        for (int i = 7; i >= 1; i--){
            if (board.getSeeds(side, i) != 0){
                if (move == i)
                    result = 1;
                break;
            }
        }
        return result;
    }
}